landingai/vision_pipeline.py

```diff
@@ -0,0 +1,338 @@
+        """Open an a window and display all the images.
+        Parameters
+        ----------
+        image_src: if empty the source image will be displayed. Otherwise the image will be selected from `other_images`
+        """
+        # TODO: Should show be a end leaf?
+        for frame in self.frames:
+            if image_src == "":
+                frame.image.show()
+            else:
+                frame.other_images[image_src].show()
+
+        # # TODO: Implement image stacking when we have multiple frames (https://answers.opencv.org/question/175912/how-to-display-multiple-images-in-one-window/)
+        # """Open an OpenCV window and display all the images. This call will stop the execution until a key is pressed.
+        # Parameters
+        # ----------
+        # image_src: if empty the source image will be displayed. Otherwise the image will be selected from `other_images`
+        # """
+        # # OpenCV is full of issues when it comes to displaying windows (see https://stackoverflow.com/questions/6116564/destroywindow-does-not-close-window-on-mac-using-python-and-opencv)
+        # cv2.namedWindow("image")
+        # cv2.startWindowThread()
+        # if image_src == "":
+        #     img = cv2.cvtColor(np.asarray(self.frames[0].image), cv2.COLOR_BGR2RGB)
+        # else:
+        #     img = cv2.cvtColor(np.asarray(self.frames[0].other_images[image_src]), cv2.COLOR_BGR2RGB)
+        # cv2.imshow("Landing AI - Press any key to exit", img)
+        # cv2.waitKey(0) # close window when a key press is detected
+        # cv2.waitKey(1)
+        # cv2.destroyWindow('image')
+        # for i in range (1,5):
+        #     cv2.waitKey(1)
+
+        return self
+
+    def apply(self, function: Callable[[Frame], Frame] = lambda f: f) -> "FrameSet":
+        """Apply a function to all frames
+
+        Parameters
+        ----------
+        function: lambda function that takes individual frames and returned an updated frame
+        """
+        for i in range(len(self.frames)):
+            self.frames[i] = function(self.frames[i])
+        return self
+
+    def filter(self, function: Callable[[Frame], bool] = lambda f: True) -> "FrameSet":
+        """Evaluate a function on every frame and keep or remove
+
+        Parameters
+        ----------
+        function : lambda function that gets invoked on every Frame. If it returns False, the Frame will be deleted
+        """
+        for i in reversed(
+            range(0, len(self.frames))
+        ):  # Traverse in reverse so we can delete
+            if not function(self.frames[i]):
+                self.frames.pop(i)
+        return self
+
+
+# openCV's default VideoCapture cannot drop frames so if the CPU is overloaded the stream will tart to lag behind realtime.
+# This class creates a treaded capture implementation that can stay up to date wit the stream and decodes frames only on demand
+class NetworkedCamera(BaseModel):
```

Comment:
just a note: i'm thinking maybe later we can define a dedicated "image source" module, or a "camera" module for better separation of concern.

Reply:
Agree.. this is a start